saleor/payment/tasks.py

```diff
@@ -30,131 +31,143 @@ def checkouts_with_funds_to_release():
-            )
+            with transaction.atomic():
+                TransactionEvent.objects.bulk_create(
+                    [event for _tr, event in transactions_with_cancel_request_events]
+                    + [event for _tr, event in transactions_with_charge_request_events]
+                )
+                # Mark transactions as not refundable to avoid multiple refund requests
+                transactions.update(last_refund_success=False)
+
             manager = get_plugins_manager(allow_replica=True)
-            for transaction, event in transactions_with_cancel_request_events:
-                channel_slug = channel_map[transaction.channel_id].slug
+            for transaction_item, event in transactions_with_cancel_request_events:
+                channel = checkout_id_to_channel[transaction_item.checkout_id]
                 try:
                     request_cancelation_action(
                         request_event=event,
                         cancel_value=event.amount_value,
                         action=TransactionAction.CANCEL,
-                        channel_slug=channel_slug,
+                        channel_slug=channel.slug,
                         user=None,
                         app=None,
-                        transaction=transaction,
+                        transaction=transaction_item,
                         manager=manager,
                     )
                 except PaymentError as e:
                     logger.warning(
                         "Unable to cancel transaction %s. %s",
-                        transaction.token,
+                        transaction_item.token,
                         str(e),
                     )
-            for transaction, event in transactions_with_charge_request_events:
-                channel_slug = channel_map[transaction.channel_id].slug
+            for transaction_item, event in transactions_with_charge_request_events:
+                channel = checkout_id_to_channel[transaction_item.checkout_id]
                 try:
                     request_refund_action(
                         request_event=event,
                         refund_value=event.amount_value,
-                        channel_slug=channel_slug,
+                        channel_slug=channel.slug,
                         user=None,
                         app=None,
-                        transaction=transaction,
+                        transaction=transaction_item,
                         manager=manager,
                     )
                 except PaymentError as e:
                     logger.warning(
                         "Unable to refund transaction %s. %s",
-                        transaction.token,
+                        transaction_item.token,
                         str(e),
                     )
+        else:
+            logger.warning("No transactions to release funds.")
```

Comment:
Maybe worth to be more explicit here 🤔  For example, In case of having 30 transactions, where `transactions_with_charge_request_events` is empty and we will have `transactions_with_cancel_request_events` single elemenent, we won't see error msg. But for some reasons, we had 30 fetched transactions, and in "standard" scenario all of them should have a charge or cancel event.